### DISCUSSION TITLE
C# implicitly change nullability of property

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#2219] [Language Ideas]
This is a proposed improvement of the null checking of properties in C#.
I have a nullable property in the base class (see SearchParameters.Url in the example code). The constructor of the child class (SuccessAnalysisParameters) ensures this property is not null.

When using the child class, I didn't expect a warning to tell me this property can be null. This is a false positive.

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/414634/c-implicitly-change-nullability-of-property.html
VSTS ticketId: 756307_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by jrmoreno1** [comment]
Is it a read only property?  If not I would think that the nullability analysis is correct.

**Comment by scalablecory** [comment]
@jrmoreno1 I think the ask is to "amend" a property of the base class to indicate that, when accessed via the derived class, there is a guarantee of non-null.

I think this can already be accomplished sufficiently, though:

```
class Base
{
    public string? Foo { get; }

    public Base(string? foo)
    {
        this.Foo = foo;
    }
}

class Derived : Base
{
    public new string Foo => base.Foo!;

    public Derived(string foo) : base(foo)
    {
    }
}
```

**Comment by 333fred** [comment]
We discussed this in LDM earlier this year and decided to support this scenario, but it has not yet been implemented. See https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-01-07.md#discussion and the bugs linked from those notes.
